daemon/cluster/executor/container/controller.go

```diff
@@ -153,10 +154,40 @@ func (r *controller) Wait(pctx context.Context) error {
 	ctx, cancel := context.WithCancel(pctx)
 	defer cancel()
 
-	err := r.adapter.wait(ctx)
-	if ctx.Err() != nil {
+	// error chan for wait
+	waitErrCh := make(chan error)
+	// error chan for health check
+	healthErrCh := make(chan error)
+
+	runner := func(f func(ctx context.Context) error, errCh chan<- error) {
+		err := f(ctx)
+		select {
+		case <-r.closed:
+		case <-ctx.Done():
+		case errCh <- err:
+		}
+	}
+
+	// run wait and checkHealth
+	go runner(r.adapter.wait, waitErrCh)
+	go runner(r.checkHealth, healthErrCh)
```

Comment:
I'm not sure what problems appeared. I think if our only action for `unhealthy` event is to shutdown container we should wait here until the shutdown has actually happened and return an error that is a combination of the wait error together with "unhealthy" cause.

If you check exit code just on container then this is not handled here. We always set the exit code but in this case, we don't seem to be setting it as a task failure reason.